### PR TITLE
8316587: Use ArraysSupport.vectorizedHashCode in Utf8EntryImpl

### PR DESCRIPTION
**Pull request by liach**
Like #12077, this uses JDK's internal utilities to speed up ASCII reading in Class files, where most identifiers, from conventions to attribute names, are ASCII. See the JBS issue for more in-depth explanations.

Before: (Master)
```
Benchmark                         Mode  Cnt    Score   Error  Units
ReadMetadata.jdkReadMemberNames  thrpt    4  167.623 ± 8.522  ops/s
```

After: (This patch, first revision)
```
Benchmark                         Mode  Cnt    Score   Error  Units
ReadMetadata.jdkReadMemberNames  thrpt    4  175.908 ± 4.766  ops/s
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316587](https://bugs.openjdk.org/browse/JDK-8316587): Use ArraysSupport.vectorizedHashCode in Utf8EntryImpl (**Enhancement** - P4)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)
 * [Claes Redestad](https://openjdk.org/census#redestad) (@cl4es - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15837/head:pull/15837` \
`$ git checkout pull/15837`

Update a local copy of the PR: \
`$ git checkout pull/15837` \
`$ git pull https://git.openjdk.org/jdk.git pull/15837/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15837`

View PR using the GUI difftool: \
`$ git pr show -t 15837`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15837.diff">https://git.openjdk.org/jdk/pull/15837.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15837#issuecomment-1727332377)

### CODE REVIEW

**Comment by liach** [comment]
@asotona Since this is your area, can you take a look?
@cl4es Can you review this as well since you are more professional in string optimizations, and can you see if you have similar performance improvements?

**Comment by cl4es** [comment]
I can verify the improvement locally in your added micro:
```
Name                            Cnt         Base    Error          Test    Error   Unit   Diff%
ReadMetadata.jdkReadMemberNames   4      108.234 ±  1.089       114.953 ±  0.709  ops/s    6.2% (p = 0.000*)
```
I don't have domain expertise w.r.t the classfile API to value how much this will benefit typical use-cases, but it seem like a straightforward enough improvement to me. (At least the `vectorizedHashCode` has the nice property that it's always faster than a naive `31 * hash + v` loop, even when inputs are too small to actually benefit from SIMD instructions).

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 1]
```diff
@@ -226,34 +232,21 @@ enum State { RAW, BYTE, CHAR, STRING }
...
-                    foundHigh = true;
-                    break;
-                }
-                hash = 31 * hash + c;
-                px++;
-            }
-
-            if (!foundHigh) {
+            int singleBytes = JLA.countPositives(rawBytes, offset, rawLen);
+            int hash = ArraysSupport.vectorizedHashCode(rawBytes, offset, singleBytes, 0, ArraysSupport.T_BOOLEAN);
```
Comment (asotona) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:236]: We should keep in mind that this code is on lambdas bootstrap critical path.
Any foreign code may involve lambdas or method references and cause dead end in lambdas initialization after JDK-8294960.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:236]: This is safe: these two functions are already used in String itself (see StringLatin1) which is definitely loaded and initialized before Classfile API and its implementations are.

**Review by asotona — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java (hunk 1 above)
Comment (cl4es) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:236]: Agreed w.r.t bootstrap safety: `countPositives` and `vectorizedHashCode` are straightforward java routines used by `String` and others very early in the bootstrap sequence. The speed-up they bring come from being intrinsified by the JIT.

**Review by cl4es — APPROVED** [review]

**Comment by liach** [comment]
Can anyone sponsor this patch? This should be compatible with Class-File transition to preview as this does not change any API.

**Pull request closed without merging**